src/main/java/com/abdullah/coding/challenge/services/CabService.java

```diff
@@ -0,0 +1,114 @@
+package com.abdullah.coding.challenge.services;
+
+import com.abdullah.coding.challenge.Response;
+import com.abdullah.coding.challenge.dto.CabDTO;
+import com.abdullah.coding.challenge.entities.Cab;
+import com.abdullah.coding.challenge.enums.CabStatus;
+import com.abdullah.coding.challenge.repository.CabRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+
+import java.util.Optional;
+
+@Service
+public class CabService {
+    @Autowired
+    private CabRepository repository;
+
+    public Response<?> makeCabAvailable(int cabId) {
+        try {
+
+            Optional<Cab> cab = repository.findById(cabId);
+            if (cab.isEmpty()) {
+                return new Response<>(HttpStatus.NOT_FOUND, "No such cab found", null);
+            }
+
+            Cab cabUpdate = cab.get();
+            cabUpdate.setStatus(CabStatus.AVAILABLE.toString());
+            repository.save(cabUpdate);
+
+            return new Response<>(HttpStatus.NO_CONTENT, "Success", null);
+
+        } catch (Exception ex) {
+            return new Response<>(HttpStatus.INTERNAL_SERVER_ERROR, "Error occurred", null);
+        }
+    }
+
+    public Response<?> makeCabUnavailable(int cabId) {
+        try {
+
+            Optional<Cab> cab = repository.findById(cabId);
+            if (cab.isEmpty()) {
+                return new Response<>(HttpStatus.NOT_FOUND, "No such cab found", null);
+            }
+
+            Cab cabUpdate = cab.get();
+            cabUpdate.setStatus(CabStatus.NOT_AVAILABLE.toString());
+            repository.save(cabUpdate);
+
+            return new Response<>(HttpStatus.NO_CONTENT, "Success", null);
+
+
+        } catch (Exception ex) {
+            return new Response<>(HttpStatus.INTERNAL_SERVER_ERROR, "Error occurred", null);
+        }
+    }
+
+    public Response<Cab> getAvailableCab(String cabType) {
+        try {
+
+            Cab cab = repository.getCab(cabType, CabStatus.AVAILABLE.toString());
+            if (cab == null) {
+                return new Response<>(HttpStatus.NOT_FOUND, "No such cab available", null);
```

Comment:
1. Like in many other files, why haven't we used a constant file for messages like "No such cab available"? Why are we duplicating it everywhere?

Reply:
That's a good insight because I thought about it and it's perfectly doable. But, I did not implement in this particular project as it was not that complex so that I would group responses in a single file. I do have sideline project that are complex and have their business logic implemented correctly and efficiently. I can give a tour of those projects if required.